File: backend/src/bond_monitor/infrastructure/bonds/universe_cache.py
"""Shared in-process cache for enriched bond universes."""

from __future__ import annotations

import copy
import hashlib
import threading
import time
from dataclasses import dataclass
from typing import Literal

from bond_monitor.domain.bonds.models import BondRecord

CacheKind = Literal["universe", "screener"]

_DEFAULT_TTL_SEC = 120.0

_lock = threading.Lock()
_cache: dict[BondCacheKey, _CacheEntry] = {}
_ttl_sec = _DEFAULT_TTL_SEC


@dataclass(frozen=True)
class BondCacheKey:
    key_rate: float
    tax_rate: float
    token_fingerprint: str
    kind: CacheKind
    filter_by: str = ""
    max_days: int = 0
    min_volume_rub: float = 0.0


@dataclass
class _CacheEntry:
    bonds: list[BondRecord]
    source: str
    cached_at: float


def configure_ttl(seconds: float) -> None:
    """Set shared cache TTL (used from settings at startup / tests)."""
    global _ttl_sec
    _ttl_sec = seconds


def token_fingerprint(token: str | None) -> str:
    if not token:
        return ""
    return hashlib.sha256(token.encode()).hexdigest()[:16]


def get(key: BondCacheKey) -> tuple[list[BondRecord], str] | None:
    now = time.monotonic()
    with _lock:
        entry = _cache.get(key)
        if entry is None:
            return None
        if (now - entry.cached_at) >= _ttl_sec:
            del _cache[key]
            return None
        return _clone_bonds(entry.bonds), entry.source


def put(key: BondCacheKey, bonds: list[BondRecord], source: str) -> None:
    with _lock:
        _cache[key] = _CacheEntry(
            bonds=_clone_bonds(bonds),
            source=source,
            cached_at=time.monotonic(),
        )


def invalidate_all() -> None:
    with _lock:
        _cache.clear()


def _clone_bonds(bonds: list[BondRecord]) -> list[BondRecord]:
    return [copy.copy(bond) for bond in bonds]

File: backend/internal/infrastructure/bonds/universe_cache.go
package bonds

import (
	"crypto/sha256"
	"encoding/hex"
	"sync"
	"time"

	domainBonds "github.com/tonatos/bond-monitor/backend/internal/domain/bonds"
)

const defaultTTL = 120 * time.Second

// CacheKey identifies an enriched bond universe cache entry.
type CacheKey struct {
	KeyRate          float64
	TaxRate          float64
	TokenFingerprint string
}

type cacheEntry struct {
	bonds    []domainBonds.BondRecord
	source   string
	cachedAt time.Time
}

var (
	cacheMu sync.RWMutex
	cache   = map[CacheKey]cacheEntry{}
	ttl     = defaultTTL
)

// ConfigureTTL sets shared cache TTL.
func ConfigureTTL(d time.Duration) {
	ttl = d
}

// TokenFingerprint returns a short hash of the trading token.
func TokenFingerprint(token string) string {
	if token == "" {
		return ""
	}
	sum := sha256.Sum256([]byte(token))
	return hex.EncodeToString(sum[:8])
}

// Get returns cached bonds if fresh.
func Get(key CacheKey) ([]domainBonds.BondRecord, string, bool) {
	cacheMu.RLock()
	entry, ok := cache[key]
	cacheMu.RUnlock()
	if !ok || time.Since(entry.cachedAt) >= ttl {
		return nil, "", false
	}
	return cloneBonds(entry.bonds), entry.source, true
}

// Put stores bonds in cache.
func Put(key CacheKey, bs []domainBonds.BondRecord, source string) {
	cacheMu.Lock()
	cache[key] = cacheEntry{bonds: cloneBonds(bs), source: source, cachedAt: time.Now()}
	cacheMu.Unlock()
}

// InvalidateAll clears the RAM cache.
func InvalidateAll() {
	cacheMu.Lock()
	cache = map[CacheKey]cacheEntry{}
	cacheMu.Unlock()
}

// CloneBondRecord returns a copy safe for duration scoring.
func CloneBondRecord(b domainBonds.BondRecord) domainBonds.BondRecord {
	cp := b
	if b.ProfileScores != nil {
		cp.ProfileScores = make(map[string]float64, len(b.ProfileScores))
		for k, v := range b.ProfileScores {
			cp.ProfileScores[k] = v
		}
	}
	return cp
}

func cloneBonds(bs []domainBonds.BondRecord) []domainBonds.BondRecord {
	out := make([]domainBonds.BondRecord, len(bs))
	for i, b := range bs {
		out[i] = CloneBondRecord(b)
	}
	return out
}
